### PR TITLE
[SPARK-19067][SS] Processing-time-based timeout in MapGroupsWithState

### PR DESCRIPTION
**Pull request by tdas**
## What changes were proposed in this pull request?

When a key does not get any new data in `mapGroupsWithState`, the mapping function is never called on it. So we need a timeout feature that calls the function again in such cases, so that the user can decide whether to continue waiting or clean up (remove state, save stuff externally, etc.).
Timeouts can be either based on processing time or event time. This JIRA is for processing time, but defines the high level API design for both. The usage would look like this.
```
def stateFunction(key: K, value: Iterator[V], state: KeyedState[S]): U = {
  ...
  state.setTimeoutDuration(10000)
  ...
}

dataset					// type is Dataset[T]
  .groupByKey[K](keyingFunc)   // generates KeyValueGroupedDataset[K, T]
  .mapGroupsWithState[S, U](
     func = stateFunction, 
     timeout = KeyedStateTimeout.withProcessingTime)	// returns Dataset[U]
```

Note the following design aspects.

- The timeout type is provided as a param in mapGroupsWithState as a parameter global to all the keys. This is so that the planner knows this at planning time, and accordingly optimize the execution based on whether to saves extra info in state or not (e.g. timeout durations or timestamps).

- The exact timeout duration is provided inside the function call so that it can be customized on a per key basis.

- When the timeout occurs for a key, the function is called with no values, and KeyedState.isTimingOut() set to true.

- The timeout is reset for key every time the function is called on the key, that is, when the key has new data, or the key has timed out. So the user has to set the timeout duration everytime the function is called, otherwise there will not be any timeout set.

Guarantees provided on timeout of key, when timeout duration is D ms:
- Timeout will never be called before real clock time has advanced by D ms
- Timeout will be called eventually when there is a trigger with any data in it (i.e. after D ms). So there is a no strict upper bound on when the timeout would occur. For example, if there is no data in the stream (for any key) for a while, then the timeout will not be hit.

Implementation details:
- Added new param to `mapGroupsWithState` for timeout
- Added new method to `StateStore` to filter data based on timeout timestamp
- Changed the internal map type of `HDFSBackedStateStore` from Java's `HashMap` to `ConcurrentHashMap` as the latter allows weakly-consistent fail-safe iterators on the map data. See comments in code for more details.
- Refactored logic of `MapGroupsWithStateExec` to 
  - Save timeout info to state store for each key that has data.
  - Then, filter states that should be timed out based on the current batch processing timestamp.
- Moved KeyedState for `o.a.s.sql` to `o.a.s.sql.streaming`. I remember that this was a feedback in the MapGroupsWithState PR that I had forgotten to address. 

## How was this patch tested?
New unit tests in 
- MapGroupsWithStateSuite for timeouts.
- StateStoreSuite for new APIs in StateStore.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/KeyValueGroupedDataset.scala [hunk 1]
```diff
@@ -230,16 +231,19 @@ class KeyValueGroupedDataset[K, V] private[sql](
    *
    * @tparam S The type of the user-defined state. Must be encodable to Spark SQL types.
    * @tparam U The type of the output objects. Must be encodable to Spark SQL types.
+   * @param func Function to be called on every group.
    *
-   * See [[Encoder]] for more details on what types are encodable to Spark SQL.
+   *             See [[Encoder]] for more details on what types are encodable to Spark SQL.
```
Comment (amitsela) [inline sql/core/src/main/scala/org/apache/spark/sql/KeyValueGroupedDataset.scala]: Just going through the code out of interest, and noticed this "chunk" of whitespace which I assume is accidental but I thought I'd note.

File: sql/core/src/main/scala/org/apache/spark/sql/streaming/KeyedState.scala [hunk 2]
```diff
@@ -61,25 +65,49 @@ import org.apache.spark.sql.catalyst.plans.logical.LogicalKeyedState
  *  - After that, if `update(newState)` is called, then `exists()` will again return `true`,
  *    `get()` and `getOption()`will return the updated value.
  *
+ * Important points to note about using `KeyedStateTimeout`.
+ *  - The timeout type is a global param across all the keys (set as `timeout` param in
+ *    `[map|flatMap]GroupsWithState`, but the exact timeout duration is configurable per key
+ *    (by calling `setTimeout...()` in `KeyedState`).
+ *  - When the timeout occurs for a key, the function is called with no values, and
+ *    `KeyedState.isTimingOut()` set to true.
+ *  - The timeout is reset for key every time the function is called on the key, that is,
```
Comment (amitsela) [inline sql/core/src/main/scala/org/apache/spark/sql/streaming/KeyedState.scala]: If a timeout was previously set, and an update (new input value for a key) comes, the user would still have to re-set the timeout ? why not leave it as it was ?
I guess that even with this implementation, this could still be sugar-coated by doing something like:
```
if (user not set timeout)
  timeout = prevTimeout
``` 

no ?

File: sql/core/src/main/scala/org/apache/spark/sql/streaming/KeyedState.scala [hunk 3]
```diff
@@ -61,25 +65,49 @@ import org.apache.spark.sql.catalyst.plans.logical.LogicalKeyedState
...
+ *    when the key has new data, or the key has timed out. So the user has to set the timeout
+ *    duration every time the function is called, otherwise there will not be any timeout set.
+ *  - Guarantees provided on processing-time-based timeout of key, when timeout duration is D ms:
+ *    - Timeout will never be called before real clock time has advanced by D ms
+ *    - Timeout will be called eventually when there is a trigger with any data in it
```
Comment (amitsela) [inline sql/core/src/main/scala/org/apache/spark/sql/streaming/KeyedState.scala]: I understand it's expensive to "scan-through" the state, but why not enable it periodically ? it could be useful for sparse streams.
Something a-la-updateStateByKey, but only if no data already triggered the state store for a configured time period.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/MapGroupsWithStateSuite.scala [hunk 4]
```diff
@@ -83,6 +88,34 @@ class MapGroupsWithStateSuite extends StateStoreMetricsTest with BeforeAndAfterA
...
+    assert(state.getTimeoutTimestamp === TIMEOUT_TIMESTAMP_NOT_SET)
+    intercept[UnsupportedOperationException] {
+      state.setTimeoutDuration(1000)
+    }
+    intercept[UnsupportedOperationException] {
+      state.setTimeoutDuration("1 day")
+    }
+    assert(state.getTimeoutTimestamp === TIMEOUT_TIMESTAMP_NOT_SET)
+
+    state = new KeyedStateImpl[String](None, 1000, isTimeoutEnabled = true, isTimingOut = false)
```
Comment (amitsela) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/MapGroupsWithStateSuite.scala]: Judging by the tests below, it seems that timeout timestamps are `timeout + batchInterval`, which makes total sense but it might be worth adding to documentation that time-based triggers are affected by "system-latency" or something.. I believe this is true for most streaming frameworks anyway, but this "disclaimer" might avoid user confusion as he might expect his output after 200msec while internal batch interval is 250msec.
I'm only noting this because It caught me off-guard, until i saw the 1000msec batchInterval passed to `KeyedStateImpl` constructor.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/object.scala [hunk 5]
```diff
@@ -317,6 +317,9 @@ case class MapGroups(
 /** Internal class representing State */
 trait LogicalKeyedState[S]
 
+case object NoTimeout extends KeyedStateTimeout
```
Comment (tdas) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/object.scala]: Add docs

File: sql/core/src/main/scala/org/apache/spark/sql/streaming/KeyedState.scala (hunk 2 above)
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/streaming/KeyedState.scala]: its intuitive to ensure there is always a timeout. no matter what updates you make to the state and what you return, you can always call `keyedState.setTimeoutDuration(...)`.

File: sql/core/src/main/scala/org/apache/spark/sql/streaming/KeyedState.scala (hunk 3 above)
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/streaming/KeyedState.scala]: Note that there is a fundamental difference between Spark Streaming and Structured Streaming. Unlike the former, Structured Streaming queries will not execute any batch if there is no new data. So as long as there is some data and batch kicks off, timeouts will be processed.

I think "a trigger with any data" is confusing. A trigger by definition has data. I will rewrite. Thank for pointing it out though.

File: sql/core/src/main/scala/org/apache/spark/sql/KeyValueGroupedDataset.scala (hunk 1 above)
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/KeyValueGroupedDataset.scala]: yeah. will take care of those. I have been adding more tests and restructuring the code to make it more testable.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/MapGroupsWithStateSuite.scala (hunk 4 above)
Comment (tdas) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/MapGroupsWithStateSuite.scala]: I will reword it.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/statefulOperators.scala [hunk 6]
```diff
@@ -256,94 +257,6 @@ case class StateStoreSaveExec(
   override def outputPartitioning: Partitioning = child.outputPartitioning
 }
 
-
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/statefulOperators.scala]: Moved to a separate file.

File: sql/core/src/main/scala/org/apache/spark/sql/streaming/KeyedState.scala (hunk 3 above)
Comment (amitsela) [inline sql/core/src/main/scala/org/apache/spark/sql/streaming/KeyedState.scala]: Using a `ProcessingTime` timeout could be confusing this way - a user would expect to fire on timeout (+ system latency). Adding system latency, wether it's batch-long, or 10Xbatch-long (like with checkpointing in Spark 1.x for example) is still something pre-set that could also be configurable, giving users the control to do the trade-off between accuracy in firings vs. performance.
Using am `EventTime` timeout in the future, I assume the "clock" would be watermark based instead of wall-time, and I see two use-cases where this would matter:
1. Testing - being able to move the clock forward to end-of-time to force firing everything that still awaits the closing of windows.
2. a pipeline where there is a filter before the stateful op. such that there is data, and the watermark advances, but some of the events are dropped and don't reach the stateful operator so it will hold off firing until the "proper" data (that passes filter) comes along - this again could cause an unknown delay to firing results.

File: sql/core/src/main/scala/org/apache/spark/sql/streaming/KeyedState.scala (hunk 2 above)
Comment (amitsela) [inline sql/core/src/main/scala/org/apache/spark/sql/streaming/KeyedState.scala]: Got it.
How about adding a `KeyedState#keepTimeout()` just to allow users to avoid:
```
val priorTimeout = keyedState.getTimeoutDuration()
keyedState.setTimeoutDuration(priorTimeout)
```
and instead simply call:
```
keyedState.keepTimeoutDuration()
```
Well, I'm not so sure about it... sometimes it is better to be explicit.. but I'll persist my thoughts here anyway.

File: sql/core/src/main/scala/org/apache/spark/sql/KeyValueGroupedDataset.scala [hunk 7]
```diff
@@ -249,6 +250,43 @@ class KeyValueGroupedDataset[K, V] private[sql](
...
+   * @param timeout Timeout information for groups that do not receive data for a while
+   *
+   * See [[Encoder]] for more details on what types are encodable to Spark SQL.
+   * @since 2.2.0
+   */
+  @Experimental
+  @InterfaceStability.Evolving
+  def mapGroupsWithState[S: Encoder, U: Encoder](
+      func: (K, Iterator[V], KeyedState[S]) => U,
+      timeout: KeyedStateTimeout): Dataset[U] = {
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/KeyValueGroupedDataset.scala]: `timeoutType`?

File: sql/core/src/main/scala/org/apache/spark/sql/KeyValueGroupedDataset.scala [hunk 8]
```diff
@@ -298,12 +368,14 @@ class KeyValueGroupedDataset[K, V] private[sql](
...
    *
    * See [[Encoder]] for more details on what types are encodable to Spark SQL.
-   * @since 2.1.1
+   * @since 2.2.0
    */
   @Experimental
   @InterfaceStability.Evolving
   def flatMapGroupsWithState[S: Encoder, U: Encoder](
-      func: (K, Iterator[V], KeyedState[S]) => Iterator[U], outputMode: OutputMode): Dataset[U] = {
+      func: (K, Iterator[V], KeyedState[S]) => Iterator[U],
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/KeyValueGroupedDataset.scala]: Another option here would put the function at the end so that you could do this:

```scala
df.flatMapGroupWithState(Append) { (key, iter, state: KeyedState[Int]) =>
   ...
}
```

File: sql/core/src/main/scala/org/apache/spark/sql/streaming/KeyedState.scala [hunk 9]
```diff
@@ -61,25 +65,50 @@ import org.apache.spark.sql.catalyst.plans.logical.LogicalKeyedState
...
+ *    - Timeout will never be called before real clock time has advanced by D ms
+ *    - Timeout will be called eventually when there is a trigger in the query
+ *      (i.e. after D ms). So there is a no strict upper bound on when the timeout would occur.
+ *      For example, the trigger interval of the query will affect when the timeout is actually hit.
+ *      If there is no data in the stream (for any key) for a while, then their will not be
+ *      any trigger and timeout will not be hit until there is data.
+ *
  * Scala example of using KeyedState in `mapGroupsWithState`:
  * {{{
  * // A mapping function that maintains an integer state for string keys and returns a string.
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/streaming/KeyedState.scala]: Could you update this comment to describe the timeout behavior of the function?

File: sql/core/src/main/scala/org/apache/spark/sql/streaming/KeyedState.scala [hunk 10]
```diff
@@ -61,25 +65,50 @@ import org.apache.spark.sql.catalyst.plans.logical.LogicalKeyedState
...
+ *       state.update(newState)              // Set the new state
  *     }
+ *
  *   } else {
  *     val initialState = ...
- *     state.update(initialState)  // Set the initial state
+ *     state.update(initialState)            // Set the initial state
  *   }
- *   ... // return something
+ *   state.setTimeoutDuration("1 hour")      // Set the timeout
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/streaming/KeyedState.scala]: Does this set a timeout on a removed state?  What does that do?

File: sql/core/src/main/scala/org/apache/spark/sql/streaming/KeyedState.scala [hunk 11]
```diff
@@ -61,25 +65,50 @@ import org.apache.spark.sql.catalyst.plans.logical.LogicalKeyedState
...
+ *  - The timeout is reset for key every time the function is called on the key, that is,
+ *    when the key has new data, or the key has timed out. So the user has to set the timeout
+ *    duration every time the function is called, otherwise there will not be any timeout set.
+ *  - Guarantees provided on processing-time-based timeout of key, when timeout duration is D ms:
+ *    - Timeout will never be called before real clock time has advanced by D ms
+ *    - Timeout will be called eventually when there is a trigger in the query
+ *      (i.e. after D ms). So there is a no strict upper bound on when the timeout would occur.
+ *      For example, the trigger interval of the query will affect when the timeout is actually hit.
+ *      If there is no data in the stream (for any key) for a while, then their will not be
+ *      any trigger and timeout will not be hit until there is data.
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/streaming/KeyedState.scala]: How hard to remove this limitation?  It seems like its very hard to build reliable monitoring applications on this API unless we fix this.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FlatMapGroupsWithStateExec.scala [hunk 12]
```diff
@@ -0,0 +1,270 @@
...
+    AttributeReference("stateExists", dataType = BooleanType, nullable = false)()
+  private val stateAttributes: Seq[Attribute] = {
+    val encoderSchemaAttributes = stateEncoder.schema.toAttributes
+    if (isTimeoutEnabled) {
+      encoderSchemaAttributes :+ stateExistsAttribute :+ timestampTimeoutAttribute
+    } else encoderSchemaAttributes
+  }
+
+  import KeyedStateImpl._
+  override def outputPartitioning: Partitioning = child.outputPartitioning
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FlatMapGroupsWithStateExec.scala]: This is not true, right?  They could be outputting whatever they want.

**Review by marmbrus — CHANGES_REQUESTED** [review]
I still need to look at `FlatMapGroupsWithStateExec` closely, but its looking good!  Just one major comment.

File: sql/core/src/main/scala/org/apache/spark/sql/KeyValueGroupedDataset.scala (hunk 7 above)
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/KeyValueGroupedDataset.scala]: Then should I change `KeyedStateTimeout` to `KeyedStateTimeoutType`?

File: sql/core/src/main/scala/org/apache/spark/sql/KeyValueGroupedDataset.scala (hunk 8 above)
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/KeyValueGroupedDataset.scala]: Do we do curried functions in public API? Are there known precedents?

File: sql/core/src/main/scala/org/apache/spark/sql/streaming/KeyedState.scala (hunk 11 above)
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/streaming/KeyedState.scala]: Shouldnt be tough. We will have to occasionally and periodically run a trigger with an empty batch DF. I think that should be a separate PR as that touch the StreamExecution, and we need figure out what the policy and the APIs to specify the policy should be.

File: sql/core/src/main/scala/org/apache/spark/sql/streaming/KeyedState.scala (hunk 10 above)
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/streaming/KeyedState.scala]: It does not. Once state is remove, timeouts are disabled. I will add this to the docs.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FlatMapGroupsWithStateExec.scala (hunk 12 above)
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FlatMapGroupsWithStateExec.scala]: Good catch! Fixed it.

File: sql/core/src/main/scala/org/apache/spark/sql/KeyValueGroupedDataset.scala (hunk 7 above)
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/KeyValueGroupedDataset.scala]: Up to you, that part is not confusing to me.  Calling this a `timeout` though sounds like it will cause individual records to timeout.  Could also be `timeoutConfig` or something.

File: sql/core/src/main/scala/org/apache/spark/sql/streaming/KeyedState.scala (hunk 11 above)
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/streaming/KeyedState.scala]: I think that you just find the nearest timeout with an accumulator and set the trigger duration to the min of the default and that nearest timeout.

File: sql/core/src/main/scala/org/apache/spark/sql/streaming/KeyedState.scala (hunk 10 above)
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/streaming/KeyedState.scala]: Hmm, I wonder if we should throw an exception here.  I was pretty confused what this function would do just looking at it.

**Review by zsxwing — COMMENTED** [review]
Looks pretty good Just some nits.

File: sql/core/src/main/scala/org/apache/spark/sql/KeyValueGroupedDataset.scala [hunk 13]
```diff
@@ -249,6 +250,43 @@ class KeyValueGroupedDataset[K, V] private[sql](
...
+   * state. The result Dataset will represent the objects returned by the function.
+   * For a static batch Dataset, the function will be invoked once per group. For a streaming
+   * Dataset, the function will be invoked for each group repeatedly in every trigger, and
+   * updates to each group's state will be saved across invocations.
+   * See [[org.apache.spark.sql.streaming.KeyedState]] for more details.
+   *
+   * @tparam S The type of the user-defined state. Must be encodable to Spark SQL types.
+   * @tparam U The type of the output objects. Must be encodable to Spark SQL types.
+   * @param func Function to be called on every group.
+   * @param timeout Timeout information for groups that do not receive data for a while
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/KeyValueGroupedDataset.scala]: added to the main explanation in KeyedState

File: sql/core/src/main/scala/org/apache/spark/sql/KeyValueGroupedDataset.scala [hunk 14]
```diff
@@ -284,6 +322,38 @@ class KeyValueGroupedDataset[K, V] private[sql](
...
+   * Dataset, the function will be invoked for each group repeatedly in every trigger, and
+   * updates to each group's state will be saved across invocations.
+   * See [[KeyedState]] for more details.
+   *
+   * @tparam S The type of the user-defined state. Must be encodable to Spark SQL types.
+   * @tparam U The type of the output objects. Must be encodable to Spark SQL types.
+   * @param func          Function to be called on every group.
+   * @param stateEncoder  Encoder for the state type.
+   * @param outputEncoder Encoder for the output type.
+   * @param timeout Timeout information for groups that do not receive data for a while
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/KeyValueGroupedDataset.scala]: response same as above.

File: sql/core/src/main/scala/org/apache/spark/sql/KeyValueGroupedDataset.scala [hunk 15]
```diff
@@ -298,12 +368,14 @@ class KeyValueGroupedDataset[K, V] private[sql](
...
-   * @since 2.1.1
+   * @since 2.2.0
    */
   @Experimental
   @InterfaceStability.Evolving
   def flatMapGroupsWithState[S: Encoder, U: Encoder](
-      func: (K, Iterator[V], KeyedState[S]) => Iterator[U], outputMode: OutputMode): Dataset[U] = {
+      func: (K, Iterator[V], KeyedState[S]) => Iterator[U],
+      outputMode: OutputMode,
+      timeout: KeyedStateTimeout): Dataset[U] = {
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/KeyValueGroupedDataset.scala]: added

File: sql/core/src/main/scala/org/apache/spark/sql/KeyValueGroupedDataset.scala [hunk 16]
```diff
@@ -361,18 +435,20 @@ class KeyValueGroupedDataset[K, V] private[sql](
...
    */
   @Experimental
   @InterfaceStability.Evolving
   def flatMapGroupsWithState[S, U](
       func: FlatMapGroupsWithStateFunction[K, V, S, U],
       outputMode: OutputMode,
       stateEncoder: Encoder[S],
-      outputEncoder: Encoder[U]): Dataset[U] = {
+      outputEncoder: Encoder[U],
+      timeout: KeyedStateTimeout): Dataset[U] = {
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/KeyValueGroupedDataset.scala]: added

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStore.scala [hunk 17]
```diff
@@ -50,6 +50,8 @@ trait StateStore {
   /** Get the current value of a key. */
   def get(key: UnsafeRow): Option[UnsafeRow]
 
+  def filter(condition: (UnsafeRow, UnsafeRow) => Boolean): Iterator[(UnsafeRow, UnsafeRow)]
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStore.scala]: No, the caller can simultaneous remove any key in the returned iterator. I added this information in the docs.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FlatMapGroupsWithStateExec.scala [hunk 18]
```diff
@@ -0,0 +1,270 @@
...
+      sqlContext.sessionState,
+      Some(sqlContext.streams.stateStoreCoordinator)) { case (store, iterator) =>
+        val updater = new StateStoreUpdater(store)
+
+        // Generate a iterator that returns the rows grouped by the grouping function
+        // Note that this code ensures that the filtering for timeout occurs only after
+        // all the data has been processed. This is to ensure that the timeout information of all
+        // the keys with data is updated before they are processed for timeouts.
+        val outputIterator =
+          Seq(
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FlatMapGroupsWithStateExec.scala]: Even if ++ is lazy, the `updateStateForTimedOutKeys` function will be called greedily in your way to return the iterator. That is what I am trying to prevent, because I dont want the `stateStore.filter` to be called before all the keys with data have been updated.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/KeyedStateImpl.scala [hunk 19]
```diff
@@ -60,6 +82,45 @@ private[sql] class KeyedStateImpl[S](optionalValue: Option[S]) extends KeyedStat
...
+      duration
+    } else {
+      "interval " + duration
+    }
+    val cal = CalendarInterval.fromString(intervalString)
+    if (cal == null) {
+      throw new IllegalArgumentException(
+        s"The provided duration ($duration) is not valid.")
+    }
+    val delayMs = {
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/KeyedStateImpl.scala]: Added checks and tests.

File: sql/core/src/main/scala/org/apache/spark/sql/streaming/KeyedState.scala [hunk 20]
```diff
@@ -92,27 +121,33 @@ import org.apache.spark.sql.catalyst.plans.logical.LogicalKeyedState
  *
  *      @Override
  *      public String call(String key, Iterator<Integer> value, KeyedState<Integer> state) {
- *        if (state.exists()) {
+ *        if (state.isTimingOut) {           // If called when timing out, remove the state
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/streaming/KeyedState.scala]: done.

File: sql/catalyst/src/main/java/org/apache/spark/sql/streaming/KeyedStateTimeout.java [hunk 21]
```diff
@@ -0,0 +1,45 @@
...
+/**
+ * Represents the type of timeouts possible for the Dataset operations
+ * `mapGroupsWithState` and `flatMapGroupsWithState`. See documentation on
+ * `KeyedState` for more details.
+ *
+ * @since 2.2.0
+ */
+@Experimental
+@InterfaceStability.Evolving
+public interface KeyedStateTimeout {
```
Comment (tdas) [inline sql/catalyst/src/main/java/org/apache/spark/sql/streaming/KeyedStateTimeout.java]: changed.

File: sql/core/src/main/scala/org/apache/spark/sql/streaming/KeyedState.scala (hunk 10 above)
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/streaming/KeyedState.scala]: added exception

File: sql/core/src/main/scala/org/apache/spark/sql/streaming/KeyedState.scala (hunk 9 above)
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/streaming/KeyedState.scala]: added

File: sql/core/src/main/scala/org/apache/spark/sql/streaming/KeyedState.scala (hunk 11 above)
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/streaming/KeyedState.scala]: We can do this in a future PR.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FlatMapGroupsWithStateExec.scala (hunk 18 above)
Comment (zsxwing) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FlatMapGroupsWithStateExec.scala]: The parameter in `Iterator.++` is call-by-name: `def ++[B >: A](that: => GenTraversableOnce[B]): Iterator[B]`

**Comment by tdas** [comment]
Aah. Let me test with that then. At least if this breaks there should be a
test that fails.

**Comment by zsxwing** [comment]
LGTM.

**Comment by tdas** [comment]
I am merging this to master in the interest of time. If there are comments, I will address them in a follow up PR.

**Comment by huoyongpeng** [comment]
This patch has not merged to the master yet, right?
